src/code/z_fbdemo.c

```diff
@@ -1,20 +1,118 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/func_800B2074.s")
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/FBFilter_InitData.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/func_800B2188.s")
+void FBFilter_Destroy(FBFilter* this) {
+    osSyncPrintf("fbdemo_cleanup(%08x)\n", this);
+    osSyncPrintf("msleep(100);\n");
+    Sleep_Msec(100);
+    if (this->unk_0C != NULL) {
+        SystemArena_FreeDebug(this->unk_0C, "../z_fbdemo.c", 180);
+        this->unk_0C = NULL;
+    }
+    if (this->vtxFrame1 != NULL) {
+        SystemArena_FreeDebug(this->vtxFrame1, "../z_fbdemo.c", 181);
+        this->vtxFrame1 = NULL;
+    }
+    if (this->vtxFrame2 != NULL) {
+        SystemArena_FreeDebug(this->vtxFrame2, "../z_fbdemo.c", 182);
+        this->vtxFrame2 = NULL;
+    }
+    if (this->gfx != NULL) {
+        SystemArena_FreeDebug(this->gfx, "../z_fbdemo.c", 183);
+        this->gfx = NULL;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/func_800B22B4.s")
+FBFilter* FBFilter_Init(FBFilter* this, s32 row, s32 col) {
+    osSyncPrintf("fbdemo_init(%08x, %d, %d)\n", this, row, col);
+    bzero(this, sizeof(*this));
+    this->frame = 0;
+    this->row = row;
+    this->col = col;
+    this->unk_0C = SystemArena_MallocDebug(((row * 8) + 8) * (col + 1), "../z_fbdemo.c", 195);
+    this->vtxFrame1 = SystemArena_MallocDebug((row + 1) * sizeof(Vtx) * (col + 1), "../z_fbdemo.c", 196);
+    this->vtxFrame2 = SystemArena_MallocDebug((row + 1) * sizeof(Vtx) * (col + 1), "../z_fbdemo.c", 197);
+    this->gfx = SystemArena_MallocDebug(((this->col * ((this->row * 9) + 1)) + 2) * sizeof(Gfx), "../z_fbdemo.c", 198);
+    if (this->unk_0C == NULL || this->vtxFrame1 == NULL || this->vtxFrame2 == NULL || this->gfx == NULL) {
+        osSyncPrintf("fbdemo_init allocation error\n");
+        if (this->unk_0C != NULL) {
+            SystemArena_FreeDebug(this->unk_0C, "../z_fbdemo.c", 202);
+            this->unk_0C = NULL;
+        }
+        if (this->vtxFrame1 != NULL) {
+            SystemArena_FreeDebug(this->vtxFrame1, "../z_fbdemo.c", 203);
+            this->vtxFrame1 = NULL;
+        }
+        if (this->vtxFrame2 != NULL) {
+            SystemArena_FreeDebug(this->vtxFrame2, "../z_fbdemo.c", 204);
+            this->vtxFrame2 = NULL;
+        }
+        if (this->gfx != NULL) {
+            SystemArena_FreeDebug(this->gfx, "../z_fbdemo.c", 205);
+            this->gfx = NULL;
+        }
+        return NULL;
+    }
+    FBFilter_InitGraphics(this);
+    FBFilter_InitData(this);
+    this->frame = 0;
+    return this;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/func_800B23E8.s")
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/FBFilter_Update.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fbdemo/func_800B23F0.s")
+void FBFilter_Draw(FBFilter* this, Gfx** gfxP) {
+    Gfx* gfx = *gfxP;
+
+    gSPDisplayList(gfx++, D_8012B000);
+    FBFilter_Update(this);
```

Comment:
I wouldn't name a function called inside `Draw` as `Update`, since that goes against the whole update/draw convention we use in the rest of the codebase.